### PR TITLE
Convert source files to TypeScript

### LINKED ISSUE #96: Rewrite code into TypeScript
Update all our `.js` files to be written as TypeScript instead, after we integrate it in #95.

### PR DESCRIPTION
**Pull request by pbelbin**
Closes #96

Which means: the src/* files have been made compilable and executable as TypeScript.

The test/* files have been left as JavaScript, aside from one test which is explicitly, a TypeScript source.  This test is transpiled into JavaScript and included in the set of tests.  It's part of the 'test' script in `package.json`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,10 +28,11 @@
 	},
 	"homepage": "https://github.com/sparksuite/waterfall-cli",
 	"dependencies": {
+		"chalk": "^3.0.0-beta.1",
 		"cli-table": "^0.3.1",
-		"colors": "^1.3.3",
 		"deepmerge": "^4.0.0",
 		"fuse.js": "^3.4.5",
+		"require-main-filename": "^2.0.0",
```
Comment (WesCossick) [inline package.json]: I don't think you're using this anywhere. Even if you are, I'd like to avoid doing so.

File: test/programs/pizza-ordering/cli/entry.js [hunk 2]
```diff
@@ -1,4 +1,6 @@
 // Initialize Waterfall CLI
+console.log('entry.js: __dirname = ' + __dirname);
```
Comment (WesCossick) [inline test/programs/pizza-ordering/cli/entry.js]: Why was this added and verbose changed to `false`?

Comment (pbelbin) [reply test/programs/pizza-ordering/cli/entry.js]: It was added for debugging purposes.  It can be removed now.  Same for the verbose: false.

File: package.json [hunk 3]
```diff
@@ -6,6 +6,7 @@
 	"files": [
 		"/dist"
 	],
+	"type": "commonjs",
```
[the remark this answers is not in the document]
Comment (pbelbin) [reply package.json]: Hey Wes, if you look here, you'll find that they talk about the `type` being `commonjs` and it having a specific meaning in terms of how the import processing is handled.  Perhaps it *should* be present after all.

[https://nodejs.org/api/esm.html#esm_code_package_json_code_code_type_code_field](https://nodejs.org/api/esm.html#esm_code_package_json_code_code_type_code_field)

[the remark this answers is not in the document]
Comment (pbelbin) [reply package.json]: At present, the tests pass *without* the type: 'commonjs'

[the remark this answers is not in the document]
Comment (pbelbin) [reply package.json]: Seems the docs say that specifying type:'commonjs' is the same as there not being any 'type' specified, so it's ok the way it is.

File: tsconfig.json [hunk 4]
```diff
@@ -12,7 +13,9 @@
 		"rootDir": "src",
 		"sourceMap": true,
 		"strict": true,
-		"target": "es6"
+		"strictNullChecks": true,
+		"target": "ES6",
+		"types": ["node"]
```
Comment (WesCossick) [inline tsconfig.json]: I think the `types` key should not be present. Otherwise, it negates all the `@types` packages we've saved as `devDependencies`. See: https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#types-typeroots-and-types

File: test/programs/pizza-ordering/cli/entry.js [hunk 5]
```diff
@@ -1,4 +1,5 @@
 // Initialize Waterfall CLI
-require('../../../../dist/index.js')({
+const cli = require('../../../../dist');
+cli.default({
```
Comment (WesCossick) [inline test/programs/pizza-ordering/cli/entry.js]: The approach here should not change. It should simply add `.default` after the require, like:

```
require('../../../../dist').default({
```

This suggestion also removes `index.js`.

File: src/utils.ts [hunk 6]
```diff
@@ -0,0 +1,759 @@
...
+}
+
+export interface ConstructorSettings {
+	app?: AppSettings;
+	arguments?: string[];
+	mainFilename: string;
+	newVersionWarning?: NewVersionWarningSettings;
+	onStart?: string | Function | null;
+	packageFilePath?: string;
+	spacing?: ConstructorSettingsSpacing;
```
Comment (WesCossick) [inline src/utils.ts]: I would eliminate a separate `ConstructorSettingsSpacing` interface and simply define this directly here:

```suggestion
	spacing?: {
		before?: number;
		after?: number;
	};
```

File: src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,759 @@
...
+interface NewVersionWarningSettings {
+	enabled: boolean;
+	installedGlobally: boolean;
+}
+
+export interface ConstructorSettings {
+	app?: AppSettings;
+	arguments?: string[];
+	mainFilename: string;
+	newVersionWarning?: NewVersionWarningSettings;
```
Comment (WesCossick) [inline src/utils.ts]: This doesn't need its own dedicated type; you can define it directly here. See suggestion below about `ConstructorSettingsSpacing `.

File: src/utils.ts [hunk 8]
```diff
@@ -0,0 +1,737 @@
...
+	after?: number;
+	before?: number;
+}
+
+interface NewVersionWarningSettings {
+	enabled: boolean;
+	installedGlobally: boolean;
+}
+
+export interface ConstructorSettings {
```
Comment (WesCossick) [inline src/utils.ts]: This isn't just the settings in the constructor; these are the settings for the entire Waterfall CLI instance. The name should reflect this.

```suggestion
export interface Settings {
```

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,4 @@
+import { cli, command, error } from './waterfall-cli';
+export default cli;
+export { command, error, cli };
```
Comment (WesCossick) [inline src/index.ts]: Why do we also export `cli` here, if it's already being exported as the default?

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
+import { cli, command, error } from './waterfall-cli';
```
Comment (WesCossick) [inline src/index.ts]: Let's add some commenting:

```suggestion
// Imports
import { cli, command, error } from './waterfall-cli';

// Exports
```

File: src/screens.ts [hunk 11]
```diff
@@ -1,12 +1,12 @@
...
-const Table = require('cli-table');
-const utils = require('./utils.js');
+import chalk from 'chalk';
+import Table from 'cli-table';
+import utils, { OrganizedArguments, Settings, CommandSpec } from './utils';
 
-// Helpful utility functions
-module.exports = function Constructor(currentSettings) {
+// Default export
+export function screens(currentSettings: Settings) {
```
Comment (WesCossick) [inline src/screens.ts]: This is being exported as the default later on.

```suggestion
function screens(currentSettings: Settings) {
```

File: src/utils.ts [hunk 12]
```diff
@@ -0,0 +1,723 @@
...
+	values: (string | number)[];
+}
+
+// Helpful utility functions
+export default function utils(currentSettings: Partial<Settings>) {
+	// Store an internal copy of the current settings
+	const settings: Settings = {
+		mainFilename: '',
+		verbose: false,
+		spacing: { after: 1, before: 1 },
```
Comment (WesCossick) [inline src/utils.ts]: Why did you add these keys?

File: src/index.ts (hunk 9 above)
Comment (pbelbin) [reply src/index.ts]: I did this intentionally, so that the default would be available also as a named export, so that clients would have some additional flexibility in how the module is consumed.

Comment (WesCossick) [reply src/index.ts]: I would prefer we export it one way only; otherwise, we'll have clients using this codebase in inconsistent ways. This would make updating the codebase more likely to produce breaking changes and make intentionally breaking changes harder to communicate. It would also make the documentation surrounding how to use this functionality more complex.

**Comment by pbelbin** [comment]
> There some places where you're using both `?:` and ` | null`. This is not a great design; we should eliminate this. This may require removing some settings from the default settings object. Now that we've defined what the settings looks like, we don't need those default settings to have every key anymore.
> 
> From what I've seen, the standard way of doing index signatures is with `[index: string]` instead of `[propName: string]`. I'd do a global find-and-replace to swap this out.

I'm looking at this now....  this is not resolved yet.

**Comment by pbelbin** [comment]
> > There some places where you're using both `?:` and ` | null`. This is not a great design; we should eliminate this. This may require removing some settings from the default settings object. Now that we've defined what the settings looks like, we don't need those default settings to have every key anymore.
> > From what I've seen, the standard way of doing index signatures is with `[index: string]` instead of `[propName: string]`. I'd do a global find-and-replace to swap this out.
> 
> I'm looking at this now.... this is not resolved yet.

From what I can see, it's aritrary.  I've seen content at typescriptlang.org use 'key' and 'propName' in code examples in their discussions on the topic ([a bit down from here](https://www.typescriptlang.org/docs/handbook/interfaces.html#excess-property-checks) and [a bit down from here](http://www.typescriptlang.org/docs/handbook/advanced-types.html#index-types)), and in the text, they also refer to this as the 'index' too.

So, if you want to standardize on 'index', I guess that's fine.

File: src/utils.ts [hunk 13]
```diff
@@ -0,0 +1,698 @@
...
+	data?: string | number;
+	[propName: string]: boolean | string | number | undefined;
+}
+
+// Interface declarations
+export interface AppSettings {
+	name: string | null;
+	packageName: string | null;
+	version: string | null;
+	//[propName: string]: string | number | null;
```
Comment (pbelbin) [inline src/utils.ts]: I was not sure whether the intention is to allow the 'AppSettings' to have additional arbitrary properties or not. Since this object is retrievable, I thought it conceivable that additional properties could be useful to clients. What say you?

Comment (WesCossick) [reply src/utils.ts]: No, only settings that are supported by Waterfall CLI should be used.

File: src/utils.ts [hunk 14]
```diff
@@ -0,0 +1,697 @@
...
+	command: string;
+	data?: string | number;
+	[index: string]: boolean | string | number | undefined;
+}
+
+// Interface declarations
+export interface AppSettings {
+	name: string | null;
+	packageName: string | null;
+	version: string | null;
```
[the remark this answers is not in the document]
Comment (WesCossick) [reply src/utils.ts]: Now that we're using TypeScript, our types should not use `| null` to indicate that a property is optional. That's what TypeScript's built-in `?:` syntax is for.

File: src/utils.ts [hunk 15]
```diff
@@ -0,0 +1,697 @@
...
+export interface CommandSpec {
+	data: {
+		accepts?: string[];
+		description: string;
+		ignoreFlagsAndOptions?: boolean;
+		required?: boolean;
+		type?: string;
+	};
+	description?: string;
+	flags: {
```
Comment (WesCossick) [inline src/utils.ts]: This should be optional.

```suggestion
	flags?: {
```

File: src/utils.ts [hunk 16]
```diff
@@ -0,0 +1,697 @@
...
+		[index: string]: {
+			cascades?: boolean;
+			description?: string;
+			shorthand?: string;
+		};
+	};
+	options: {
```
Comment (WesCossick) [inline src/utils.ts]: This should be optional.

```suggestion
	options?: {
```

File: src/utils.ts [hunk 17]
```diff
@@ -0,0 +1,697 @@
...
+export interface AppSettings {
+	name: string | null;
+	packageName: string | null;
+	version: string | null;
+}
+
+export interface CommandSpec {
+	data: {
+		accepts?: string[];
+		description: string;
```
Comment (WesCossick) [inline src/utils.ts]: You looked into this and said that the `description` is always required; however, I looked at the code on the `master` branch and found that it's optional. It will only be displayed if present.

```suggestion
		description?: string;
```

File: src/utils.ts [hunk 18]
```diff
@@ -0,0 +1,697 @@
...
+		ignoreFlagsAndOptions?: boolean;
+		required?: boolean;
+		type?: string;
```
Comment (WesCossick) [inline src/utils.ts]: We only support `integer` and `float` right now. Therefore, asking for any `string` value is not restrictive enough.

```suggestion
		type?: 'integer' | 'float';
```

File: src/utils.ts [hunk 19]
```diff
@@ -0,0 +1,697 @@
...
+		};
+	};
+	options: {
+		[index: string]: {
+			accepts?: string[];
+			cascades?: boolean;
+			description?: string;
+			required?: boolean;
+			shorthand?: string;
+			type?: string;
```
Comment (WesCossick) [inline src/utils.ts]: We only support `integer` and `float` right now. Therefore, asking for any `string` value is not restrictive enough.

```suggestion
			type?: 'integer' | 'float';
```

File: src/waterfall-cli.ts [hunk 20]
```diff
@@ -147,26 +172,34 @@ module.exports = function Constructor(customSettings) {
...
+			currentPathPrefix
+		);
 
 		// Get the command path
 		const commandPath = commandFiles.filter(path => path.match(/\.js$/))[0];
 
 		// Get spec
 		const specFilePath = commandFiles.filter(path => path.match(/\.json$/))[0];
-		let spec = {};
+		let spec: CommandSpec = {
```
Comment (WesCossick) [inline src/waterfall-cli.ts]: After you make the data's `description` optional, along with the `flags` and `options` keys, you should be able to leave this as `let spec = {};`.

File: src/utils.ts [hunk 21]
```diff
@@ -0,0 +1,697 @@
...
+
+// Interface declarations
+export interface AppSettings {
+	name: string | null;
+	packageName: string | null;
+	version: string | null;
+}
+
+export interface CommandSpec {
+	data: {
```
Comment (WesCossick) [inline src/utils.ts]: This should be optional:

```suggestion
	data?: {
```

File: src/utils.ts (hunk 17 above)
Comment (pbelbin) [reply src/utils.ts]: Hey Wes,

Look at the `master` `src/utils.js` around line 362 - my take is that the idea here is that if the supplied arguments, after having been organized, results in there being 'data', and, then, if either the spec's data is entirely null, or, the spec's data.description is null, then the code proceeds to generate an error message.  So, I think this means (unless it's a bug): if you have a spec.data, it *must* have a description (unless the object of the app is to endlessly produce usage error messages).

Additionally, if you look at `master`'s `src/screen.js` around line 100, it requires the presence of both the mergedSpec.data and the mergedSpec.data.description before it decides that the app 'acceptsData', again reinforcing the notion that within the `CommandSpec`, that if you have a `data` object, it requires a `description` property for it to be considered valid.

Of course, this is what it's doing now.  It could be changed :)

Let me know what you think.

Comment (WesCossick) [reply src/utils.ts]: Interesting, there does appear to be an inconsistency. On `screens.js` [line 238](https://github.com/sparksuite/waterfall-cli/blob/master/src/screens.js#L238), it treats the description as optional, which is what I was looking at.

But the other two places you mentioned do treat it as required. We'll, therefore, need to decide what's better going forward.

I'm inclined to believe that it's better as optional; this would match the descriptions for flags, options, and commands. Plus it gives the engineer using Waterfall CLI control over whether or not they think it's necessary.

We should create a separate PR to change this. I can probably do that pretty quickly, but I'll wait to merge it to see what you think.

**Comment by pbelbin** [comment]
> Let's go ahead and create a `types.ts` file that contains any types that are used across more than one file. That would include types like `Settings`, `AppSettings`, `CommandSpec`, etc. Then, any file that needs a shared type can import it from `types.ts`.

done.

File: src/screens.ts [hunk 22]
```diff
@@ -1,10 +1,11 @@
...
-const utils = require('./utils');
+import chalk from 'chalk';
+import Table from 'cli-table';
+import { Settings } from './types';
+import utils from './utils';
 
-// Helpful utility functions
-module.exports = function Constructor(currentSettings) {
+// Default export
+function screens(currentSettings: Settings) {
```
Comment (WesCossick) [inline src/screens.ts]: This would be more consistent with other files...

```suggestion
export default function screens(currentSettings: Settings) {
```

You'll need to remove the `export default screens;` line down below, too.

File: src/screens.ts [hunk 23]
```diff
@@ -20,8 +21,8 @@ module.exports = function Constructor(currentSettings) {
 				outputString += chalk.bold(`${settings.app.name}: `);
 			}
 
-			// Add version and a newline
-			outputString += `${settings.app.version}\n`;
+			// Add version (if we have one) and a newline
+			outputString += `${settings.app.version ? settings.app.version : ''}\n`;
```
Comment (WesCossick) [inline src/screens.ts]: Slightly more shorthand way of doing this:

```suggestion
			outputString += `${settings.app.version || ''}\n`;
```
